src/lib.rs

```diff
@@ -61,82 +83,116 @@ impl Ficon {
         }?;
 
         let config = fs::read_to_string(&config_path)
-            .with_context(|_| format!("Config file is missing: {}", &config_path))?;
+            .with_context(|_| format!("Config file is missing: '{}'", &config_path))?;
+
+        let config: Config = toml::from_str(&config).with_context(|_| {
+            format!(
+                "Error while parsing configuration file at '{}'",
+                config_path
+            )
+        })?;
+
+        Ok(Ficon {
+            option,
+            validated_config: <_ as TryInto<ValidatedConfig>>::try_into(config).with_context(
+                |_| format!("Validation of configuration at '{}' failed", config_path),
+            )?,
+        })
+    }
+
+    pub fn target_dir(&self) -> &Path {
+        &self.option.path
+    }
 
-        let config: Config = toml::from_str(&config)
-            .with_context(|_| "Error while parsing configuration file")?;
+    pub fn check(&mut self, path: &Path) -> Result<bool, Error> {
+        let convention_regex = self.validated_config.convention_for(path)?;
+        let file_name = filename_of(path)
+            // ignore multiple extension by default
+            // TODO: make this configurable
+            .split('.')
+            .next()
+            .unwrap_or("");
 
-        Ok(Ficon { option, config })
+        Ok(convention_regex.is_match(file_name))
     }
+}
 
-    pub fn target_dir(&self) -> &Path {
-        return self.option.path.as_ref();
+impl TryFrom<Config> for ValidatedConfig {
+    type Error = Error;
+
+    fn try_from(value: Config) -> Result<ValidatedConfig, Error> {
+        Ok(ValidatedConfig {
+            default_convention: Self::new_regex_for_convention(&value.default.convention)?,
+            patterns: match value.for_patterns {
+                Some(mut pattern_configs) => pattern_configs
+                    .drain(..)
+                    .map(|conf| {
+                        Pattern::new(&conf.pattern)
+                            .with_context(|_| format!("Failed to parse pattern '{}'", conf.pattern))
+                            .map(|pattern| ValidatedSubConfig {
+                                convention: conf.convention,
+                                convention_regex: None,
+                                pattern,
+                            })
+                    })
+                    .collect::<Result<_, _>>()?,
```

Comment:
This is one of the best tricks in the book. Collect into a `Result<_, _>`, which is useful if your iterators produces `Result<T, _>`. That way, if there is an error, you will get it into the top level. You get a pristine `Vec<T>`.